File: Assignments/01_Basics/Kinship.py
tupleList1 = [
    ("parent", "Hasib", "Rakib"),
    ("parent", "Rakib", "Sohel"),
    ("parent", "Rakib", "Rebeka"),
    ("parent", "Rashid", "Hasib"),
    ("parent", "Rebeka", "Ayesha"),
    ("parent", "Sohel", "Fatima"),
]

maleList = ["Hasib", "Rakib", "Sohel", "Rashid"]


# Functions


def findGp():
    X = str(input("Grandchild: "))
    print("Grandparent: ", end=" ")
    i = 0
    while i <= 5:
        if (tupleList1[i][0] == "parent") & (tupleList1[i][2] == X):
            for j in range(6):
                if (tupleList1[j][0] == "parent") & (
                    tupleList1[i][1] == tupleList1[j][2]
                ):
                    print(tupleList1[j][1], end=" ")
        i = i + 1


def findBrother():
    X = str(input("Sibling: "))
    print("Brother: ", end=" ")
    i = 0
    while i <= 5:
        if (tupleList1[i][0] == "parent") & (tupleList1[i][2] == X):
            for j in range(6):
                if (
                    (tupleList1[j][0] == "parent")
                    & (tupleList1[i][1] == tupleList1[j][1])
                    & (tupleList1[i][2] != tupleList1[j][2])
                    & (tupleList1[j][2] in maleList)
                ):
                    print(tupleList1[j][2], end=" ")
        i = i + 1


def findSister():
    X = str(input("Sibling: "))
    print("Sister: ", end=" ")
    i = 0
    while i <= 5:
        if (tupleList1[i][0] == "parent") & (tupleList1[i][2] == X):
            for j in range(6):
                if (
                    (tupleList1[j][0] == "parent")
                    & (tupleList1[i][1] == tupleList1[j][1])
                    & (tupleList1[i][2] != tupleList1[j][2])
                    & (tupleList1[j][2] not in maleList)
                ):
                    print(tupleList1[j][2], end=" ")
        i = i + 1


def findUncle():
    X = str(input("Niece/Nephew: "))
    print("Uncle: ", end=" ")
    i = 0
    while i <= 5:
        if (tupleList1[i][0] == "parent") & (tupleList1[i][2] == X):
            for j in range(6):
                if (tupleList1[j][0] == "parent") & (
                    tupleList1[i][1] == tupleList1[j][2]
                ):
                    for k in range(6):
                        if (
                            (tupleList1[k][0] == "parent")
                            & (tupleList1[j][1] == tupleList1[k][1])
                            & (tupleList1[j][2] != tupleList1[k][2])
                            & (tupleList1[k][2] in maleList)
                        ):
                            print(tupleList1[k][2], end=" ")
        i = i + 1


def findAunt():
    X = str(input("Niece/Nephew: "))
    print("Aunt: ", end=" ")
    i = 0
    while i <= 5:
        if (tupleList1[i][0] == "parent") & (tupleList1[i][2] == X):
            for j in range(6):
                if (tupleList1[j][0] == "parent") & (
                    tupleList1[i][1] == tupleList1[j][2]
                ):
                    for k in range(6):
                        if (
                            (tupleList1[k][0] == "parent")
                            & (tupleList1[j][1] == tupleList1[k][1])
                            & (tupleList1[j][2] != tupleList1[k][2])
                            & (tupleList1[k][2] not in maleList)
                        ):
                            print(tupleList1[k][2], end=" ")
        i = i + 1


switch = {
    "findGp": findGp,
    "findBrother": findBrother,
    "findSister": findSister,
    "findUncle": findUncle,
    "findAunt": findAunt,
}

# Main

func = str(input("?- "))
if func in switch:
    switch[func]()
else:
    print("Invalid function")

File: Assignments/01_Basics/Kinship.pl
parent('Hasib', 'Rakib').
parent('Rakib','Sohel').
parent('Rakib','Rebeka').
parent('Rashid','Hasib').
parent('Rebeka','Ayesha').
parent('Sohel','Fatima').
male('Hasib').
male('Rakib').
male('Sohel').
male('Rashid').

grandparent(X,Z):-parent(X,Y),parent(Y,Z).
brother(X,Z):-parent(Y,X),parent(Y,Z),male(X),not(X=Z).
sister(X,Z):-parent(Y,X),parent(Y,Z),not(male(X)),not(X=Z).
uncle(X,Z):-parent(Y,Z),brother(X,Y).
aunt(X,Z):-parent(Y,Z),sister(X,Y).


findGp:-write('Grandchild: '),read(X),write('Grandparents: '),
    grandparent(Gp,X),write(Gp),tab(5),fail.
findGp.

findBrother:-write('Sibling: '),read(X),write('Brothers: '),
    brother(Bro,X),write(Bro),tab(5),fail.
findBrother.

findSister:-write('Sibling: '),read(X),write('Sisters: '),
    sister(Sis,X),write(Sis),tab(5),fail.
findSister.


findUncle:-write('Niece/Nephew: '),read(X),write('Uncles: '),
    uncle(Unc,X),write(Unc),tab(5),fail.
findUncle.

findAunt:-write('Niece/Nephew: '),read(X),write('Aunt: '),
    aunt(Ant,X),write(Ant),tab(5),fail.
findAunt.